op-supervisor/supervisor/backend/cross/cycle.go

```diff
@@ -0,0 +1,300 @@
+func gatherBlockData(d CycleCheckDeps, inTimestamp uint64, hazards map[types.ChainIndex]types.BlockSeal) (
+	map[types.ChainIndex]uint32, // logCounts
+	map[types.ChainIndex]map[uint32]*types.ExecutingMessage, // execMsgs
+	error,
+) {
+	logCounts := make(map[types.ChainIndex]uint32)
+	execMsgs := make(map[types.ChainIndex]map[uint32]*types.ExecutingMessage)
+
+	for hazardChainIndex, hazardBlock := range hazards {
+		// TODO(#11105): translate chain index to chain ID
+		hazardChainID := types.ChainIDFromUInt64(uint64(hazardChainIndex))
+		bl, logCount, msgs, err := d.OpenBlock(hazardChainID, hazardBlock.Number)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to open block: %w", err)
+		}
+		if bl != hazardBlock {
+			return nil, nil, fmt.Errorf("tried to open block %s of chain %s, but got different block %s than expected, use a reorg lock for consistency", hazardBlock, hazardChainID, bl)
+		}
+
+		// Validate executing message indices
+		if err := validateExecMsgs(logCount, msgs); err != nil {
+			return nil, nil, err
+		}
+
+		logCounts[hazardChainIndex] = logCount
+
+		// Initialize map for this chain if it has executing messages
+		if len(msgs) > 0 {
+			if _, exists := execMsgs[hazardChainIndex]; !exists {
+				execMsgs[hazardChainIndex] = make(map[uint32]*types.ExecutingMessage)
+			}
+		}
+
+		// Process executing messages - only include those at inTimestamp
+		for logIdx, msg := range msgs {
+			if msg.Timestamp == inTimestamp {
+				execMsgs[hazardChainIndex][logIdx] = msg
+			}
+		}
+	}
+
+	return logCounts, execMsgs, nil
+}
+
+// validateExecMsgs ensures all executing message log indices are valid
+func validateExecMsgs(logCount uint32, execMsgs map[uint32]*types.ExecutingMessage) error {
+	for logIdx := range execMsgs {
+		if logIdx >= logCount {
+			return fmt.Errorf("%w: log index %d >= log count %d", ErrInvalidLogIndex, logIdx, logCount)
+		}
+	}
+	return nil
+}
+
+// buildGraph constructs a dependency graph from the hazard blocks.
+func buildGraph(d CycleCheckDeps, inTimestamp uint64, hazards map[types.ChainIndex]types.BlockSeal) (*graph, error) {
+	g := &graph{
+		inDegree0:     make(map[msgKey]struct{}),
+		inDegreeNon0:  make(map[msgKey]uint32),
+		outgoingEdges: make(map[msgKey][]msgKey),
+	}
+
+	logCounts, execMsgs, err := gatherBlockData(d, inTimestamp, hazards)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add nodes for each log in the block, and add edges between sequential logs
+	for hazardChainIndex, logCount := range logCounts {
+		for i := uint32(0); i < logCount; i++ {
+			k := msgKey{
+				chainIndex: hazardChainIndex,
+				logIndex:   i,
+			}
+
+			if i == 0 {
+				// First log in block has no dependencies.=
+				g.inDegree0[k] = struct{}{}
+			} else {
+				// Add edge: prev log <> current log
+				prevKey := msgKey{
+					chainIndex: hazardChainIndex,
+					logIndex:   i - 1,
+				}
+				g.addEdge(prevKey, k)
```

Comment:
I believe this came about during a call with @protolambda when we discussed this log-oriented graph instead of the block-oriented one outlined in the original spec, but I don't recall the exact reasoning for the edges between sequential logs instead of only between executing and initiating msgs.

Trying it without these edges seems to make the cycle finding break in some of our cases so I think we need to really understand this piece.